experimental/BasisLieHighestWeight/src/ModuleData.jl

```diff
@@ -0,0 +1,92 @@
+abstract type ModuleData{C <: FieldElem, LieT <: LieAlgebraElem{C}} end
+
+# To be implemented by subtypes:
+# Mandatory:
+#   base_lie_algebra(V::MyModuleData) -> LieAlgebra
+#   highest_weight(V::MyModuleData) -> WeightLatticeElem
+#   dim(V::MyModuleData) -> ZZRingElem
+#   character(V::MyModuleData) -> Dict{WeightLatticeElem,ZZRingElem}
+
+
+mutable struct SimpleModuleData{C <: FieldElem, LieT <: LieAlgebraElem{C}} <: ModuleData{C, LieT}
+    L::LieAlgebra{C} # parent_type(LieT)
+    highest_weight::WeightLatticeElem
+
+    # The following fields are not set by default, just for caching
+    dim::ZZRingElem
+    character::Dict{WeightLatticeElem, ZZRingElem}
+
+    function SimpleModuleData(L::LieAlgebra{C}, highest_weight::WeightLatticeElem) where {C <: FieldElem}
+        return new{C, elem_type(L)}(L, highest_weight)
+    end
+end
+
+function SimpleModuleData(L::LieAlgebra, highest_weight::Vector{Int})
+    return SimpleModuleData(L, WeightLatticeElem(root_system(L), highest_weight))
+end
+
+function base_lie_algebra(V::SimpleModuleData{C, LieT}) where {C <: FieldElem, LieT <: LieAlgebraElem{C}}
+    return V.L::parent_type(LieT)
+end
+
+function highest_weight(V::SimpleModuleData)
+    return V.highest_weight
+end
+
+function dim(V::SimpleModuleData)
+    if !isdefined(V, :dim)
+        V.dim = dim_of_simple_module(ZZRingElem, base_lie_algebra(V), highest_weight(V))
+    end
+    return V.dim
+end
+
+function character(V::SimpleModuleData)
+    if !isdefined(V, :character)
+        V.character = character(ZZRingElem, base_lie_algebra(V), highest_weight(V))
+    end
+    return V.character
+end
+
+mutable struct DemazureModuleData{C <: FieldElem, LieT <: LieAlgebraElem{C}} <: ModuleData{C, LieT}
+    L::LieAlgebra{C} # parent_type(LieT)
+    highest_weight::WeightLatticeElem
+    weyl_group_elem::WeylGroupElem
+
+    # The following fields are not set by default, just for caching
+    dim::ZZRingElem
+    character::Dict{WeightLatticeElem, ZZRingElem}
+
+    function DemazureModuleData(L::LieAlgebra{C}, highest_weight::WeightLatticeElem, weyl_group_elem::WeylGroupElem) where {C <: FieldElem}
+        return new{C, elem_type(L)}(L, highest_weight, weyl_group_elem)
+    end
+end
+
+function DemazureModuleData(L::LieAlgebra, highest_weight::Vector{Int}, weyl_group_elem::Vector{Int})
+    return DemazureModuleData(L, WeightLatticeElem(root_system(L), highest_weight), WeylGroupElem(weyl_group(root_system(L)), weyl_group_elem))
+end
+
+function base_lie_algebra(V::DemazureModuleData{C, LieT}) where {C <: FieldElem, LieT <: LieAlgebraElem{C}}
+    return V.L::parent_type(LieT)
+end
+
+function highest_weight(V::DemazureModuleData)
+    return V.highest_weight
+end
+
+function character(V::DemazureModuleData)
+    if !isdefined(V, :character)
+        V.character = demazure_character(ZZRingElem, base_lie_algebra(V), highest_weight(V), V.weyl_group_elem)
```

Comment:
```suggestion
        _char = demazure_character(ZZRingElem, base_lie_algebra(V), highest_weight(V), V.weyl_group_elem)
        V.character = Dict([(w * inv(weyl_group_elem(V)) => k) for (w, k) in _char])
```
this is how I got the example working during today's call. but as we discussed afterwards, this is not the optimal place to do the twist but that's something we could discuss later